bdns/servers.go

```diff
@@ -0,0 +1,159 @@
+// addresses, and refreshes it regularly using a goroutine started by its
+// constructor.
+type dynamicProvider struct {
+	host string
+	// A map of IP addresses (Target fields in SRV records) to ports (Port fields
+	// in SRV records) associated with those addresses.
+	addrs  map[string][]uint16
+	mu     sync.RWMutex
+	cancel chan interface{}
+}
+
+var _ ServerProvider = &dynamicProvider{}
+
+// StartDynamicProvider constructs a new dynamicProvider and starts its
+// auto-update goroutine. The auto-update process queries DNS for SRV records
+// at refresh intervals and uses the resulting IP/port combos to populate the
+// list returned by Addrs. The update process ignores the Priority and Weight
+// attributes of the SRV records.
+func StartDynamicProvider(server string, refresh time.Duration) (*dynamicProvider, error) {
+	if server == "" {
+		return nil, fmt.Errorf("no DNS host provided")
+	}
+	updateCounter := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "dns_update",
+			Help: "Counter of attempts to update a dynamic provider",
+		},
+		[]string{"success"},
+	)
+	dp := dynamicProvider{
+		host:   server,
+		addrs:  make(map[string][]uint16),
+		cancel: make(chan interface{}),
+	}
+	err := dp.update()
+	if err != nil {
+		return nil, fmt.Errorf("failed to start dynamic provider: %w", err)
+	}
+
+	go func() {
+		t := time.NewTicker(refresh)
+		for {
+			select {
+			case <-t.C:
+				err := dp.update()
+				if err != nil {
+					updateCounter.With(prometheus.Labels{
+						"success": "false",
+					}).Inc()
+					continue
+				}
+				updateCounter.With(prometheus.Labels{
+					"success": "true",
+				}).Inc()
+			case <-dp.cancel:
+				return
+			}
+		}
+	}()
+
+	return &dp, nil
+}
+
+func (dp *dynamicProvider) update() error {
+	_, srvs, err := net.LookupSRV("dns", "", dp.host)
+	if err != nil {
+		return fmt.Errorf("failed to lookup SRV records for %q: %w", dp.host, err)
+	}
+	if srvs == nil || len(srvs) == 0 {
+		return fmt.Errorf("no SRV records found for %q", dp.host)
+	}
+
+	addrPorts := make(map[string][]uint16)
+	for _, srv := range srvs {
+		addrs, err := net.LookupHost(srv.Target)
+		if err != nil {
+			return fmt.Errorf("failed to resolve SRV Target %q: %w", srv.Target, err)
+		}
+		for _, addr := range addrs {
+			addrPorts[addr] = append(addrPorts[addr], srv.Port)
+		}
+	}
+
+	dp.mu.Lock()
+	dp.addrs = addrPorts
+	dp.mu.Unlock()
+	return nil
+}
+
+// Addrs returns a shuffled list of IP/port pairs, with the guarantee that no
+// two IP/port pairs will share the same IP.
+func (dp *dynamicProvider) Addrs() ([]string, error) {
+	var r []string
+	dp.mu.RLock()
+	for ip, ports := range dp.addrs {
+		port := ports[rand.Intn(len(ports))]
+		addr := fmt.Sprintf("%s:%d", ip, port)
+		r = append(r, addr)
+	}
+	dp.mu.RUnlock()
+	rand.Shuffle(len(r), func(i, j int) {
+		r[i], r[j] = r[j], r[i]
+	})
+	return r, nil
+}
+
+func (dp *dynamicProvider) Stop() {
+	close(dp.cancel)
+}
```

Comment:
This should have a doccomment, noting in particular that it doesn't wait for the update thread to actually stop, but just sends a signal to stop.